File: core/src/test/java/com/mower/domain/MowerShould.java
package com.mower.domain;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static com.mower.domain.CardinalPoint.NORTH;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class MowerShould {

  public static final String EXPECTED_MOWER_SITUATION = "1 2 N";
  private static final int COORDINATE_X = 1;
  private static final int COORDINATE_Y = 2;
  private static final CardinalPoint CARDINAL_POINT = NORTH;
  @Mock
  Command commandMocked;
  @Mock
  FaceTo faceToMocked;
  @Mock
  Coordinates coordinatesMocked;

  @Test
  void returnCoordinates() {
    assertEquals(anyCoordinates(), anyMower().coordinates());
  }

  @Test
  void executeCommand() {
    doNothing().when(commandMocked).execute(any(Mower.class));
    anyMower().executeCommand(commandMocked);
    verify(commandMocked).execute(any(Mower.class));
  }

  @Test
  void spinLeft() {
    doNothing().when(faceToMocked).spinLeft();
    anyMowerWithFaceToMocked().spinLeft();
    verify(faceToMocked).spinLeft();
  }

  @Test
  void spinRight() {
    doNothing().when(faceToMocked).spinRight();
    anyMowerWithFaceToMocked().spinRight();
    verify(faceToMocked).spinRight();
  }

  @Test
  void moveForward() {
    doNothing().when(coordinatesMocked).moveTowards(any(CardinalPoint.class));
    anyMowerWithCoordinatesMocked().moveForward();
    verify(coordinatesMocked).moveTowards(any(CardinalPoint.class));
  }

  @Test
  void returnSituation() {
    assertThat(anyMower().situation()).isEqualTo(EXPECTED_MOWER_SITUATION);
  }

  private Mower anyMower() {
    return new Mower(anyCoordinates(), anyFaceTo());
  }

  private Mower anyMowerWithFaceToMocked() {
    return new Mower(anyCoordinates(), faceToMocked);
  }

  private Mower anyMowerWithCoordinatesMocked() {
    return new Mower(coordinatesMocked, anyFaceTo());
  }

  private Coordinates anyCoordinates() {
    return new Coordinates(COORDINATE_X, COORDINATE_Y);
  }

  private FaceTo anyFaceTo() {
    return new FaceTo(CARDINAL_POINT);
  }
}

File: core/src/test/kotlin/com/mower/domain/MowerShould.kt
package com.mower.domain

import com.mower.domain.valueobjects.Coordinates
import com.mower.domain.valueobjects.FaceTo
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.mockito.Mockito.mock
import kotlin.test.assertEquals

class MowerShould {

    private val EXPECTED_MOWER_SITUATION = "1 2 N"
    private val COORDINATE_X = 1
    private val COORDINATE_Y = 2
    private val CARDINAL_POINT = CardinalPoint.NORTH

    private val commandMocked: Command = mock(Command::class.java)
    private val faceToMocked: FaceTo = mock(FaceTo::class.java)
    private val coordinatesMocked: Coordinates = mock(Coordinates::class.java)

    @Test
    fun returnCoordinates() {
        assertEquals(anyCoordinates(), anyMower().coordinates())
    }

    @Test
    fun executeCommand() {
        Mockito.doNothing().`when`(commandMocked).execute(ArgumentMatchers.any(Mower::class.java))
        anyMower().executeCommand(commandMocked)
        Mockito.verify(commandMocked).execute(ArgumentMatchers.any(Mower::class.java))
    }

    @Test
    fun spinLeft() {
        Mockito.doNothing().`when`(faceToMocked).spinLeft()
        anyMowerWithFaceToMocked().spinLeft()
        Mockito.verify(faceToMocked).spinLeft()
    }

    @Test
    fun spinRight() {
        Mockito.doNothing().`when`(faceToMocked).spinRight()
        anyMowerWithFaceToMocked().spinRight()
        Mockito.verify(faceToMocked).spinRight()
    }

    @Test
    fun moveForward() {
        Mockito.doNothing().`when`(coordinatesMocked).moveTowards(
            ArgumentMatchers.any(
                CardinalPoint::class.java
            )
        )
        anyMowerWithCoordinatesMocked().moveForward()
        Mockito.verify(coordinatesMocked).moveTowards(ArgumentMatchers.any(CardinalPoint::class.java))
    }

    @Test
    fun returnSituation() {
        assertEquals(EXPECTED_MOWER_SITUATION, anyMower().situation())
    }

    private fun anyMower(): Mower {
        return Mower(anyCoordinates(), anyFaceTo())
    }

    private fun anyMowerWithFaceToMocked(): Mower {
        return Mower(anyCoordinates(), faceToMocked)
    }

    private fun anyMowerWithCoordinatesMocked(): Mower {
        return Mower(coordinatesMocked, anyFaceTo())
    }

    private fun anyCoordinates(): Coordinates {
        return Coordinates(COORDINATE_X, COORDINATE_Y)
    }

    private fun anyFaceTo(): FaceTo {
        return FaceTo(CARDINAL_POINT)
    }
}
